### ISSUE TITLE
How to obtain UMAP coordinates and cluster ID per cell?

### ISSUE THREAD

**Issue by denvercal1234GitHub** [ebecht/infinityFlow#14]
Hi there, 

Thanks for the package.

The output of the pipeline produced FCS files and pdf of UMAP images. 

I do aim to use these FCS files in other packages, such as Spectre package to analyse the data, but is there a way I could have the UMAP coordinates and clustering IDs of each cell produced by `infinityFlow` pipeline, so I can explore the UMAP programmatically instead of just having the PDFs?

Thank you.

**Comment by ebecht** [comment]
Hi @denvercal1234GitHub 

By default you should be getting new channels in your FCS files with names "UMAP1" and "UMAP2" that correspond to (linearly-rescaled) UMAP embeddings of the backbone.

Clustering is not part of the pipeline so you would have to run that yourself.

If you want to run UMAP or clustering on the imputed data, this is not part of the pipeline at the moment. The two advices I would give you are to

1. Use the background-corrected files if you autofluorescence / unspecific binding is affecting the PE intensity (see Sup Fig 5 of the paper)
2. Exclude events with close to saturating measurements on the PE channel. If you don't know what I am talking about, you may want to color the UMAP based on raw PE intensity (across all wells in parallel). Cells with PE saturation issues should appear as distinct clumps on the UMAP (Sup Fig 14 of the paper).

Best,
Etienne

**Comment by denvercal1234GitHub** [comment]
Hi @ebecht - Thank you so much for your response. 

QUESTION 1. When you mentioned "exclude events with close to saturating measurements on the PE channel," you meant I should perform reduction on unimputed FCS (i.e., before performing inifnityFlow) in order to identify and remove these events before imputing using Infinityflow? Or, did you mean to do the infinityflow anyway then perform downstream reduction of the cells and subsequently visualize them on UMAP?

QUESTION 2. Do you recommend to filter out all events across FCS files before doing inifinityFlow that have expression above 10^6? My data was acquired using Cytek Aurora 4 lasers.

Thank you again for your help.

**Comment by ebecht** [comment]
1. I am not super familiar with spectral flow sadly, so I cannot precisely answer your questions.

For conventional flow cytometry, the levels at which detectors saturate are specified by vendors (although on our data we thought that vendors were a bit optimistic). They way we did this in our paper was to use the manufacturer thresholds/32 as a cutoff to exclude events. 

One way to verify this is to color the UMAP embedding of the backbone data by raw PE intensity. Any clump of cells with very bright PE intensity is likely caused by saturation (which causes non-correctable spillover issues in other channels and thus is seen in the UMAP embedding of the backbone data).

2. Yes I think that you should remove these events across all files

**Comment by denvercal1234GitHub** [comment]
Thanks @ebecht! For Q1, how did you come up with the number 32? 

My apologies, but the QC of the imputed data is giving me a bit of issues. Really looking forward to your QC tools (plots to identify saturation and look at prediction versus measurements on held-out events with computed AUC and r) as mentioned in issue #19.

Thanks again for your help!

**Comment by ebecht** [comment]
Hi,

The number 32 came from the analysis of the UMAP embedding on imputed data. It showed that cells corresponding to satured predictions corresponded to the the manufacturer's saturation limit/32. This may depend on the instrument.

You can check supplementary figure 15 of our paper to see how to choose that threshold for your dataset. The goal is basically to exclude those small clusters of cells with high PE intensity. These are problematic because saturation means that other backbone channels are undercompensated for PE, which means that the models see events with a specific backbone signature associated with high PE signal and end up predicting high expression for any saturating antibody on these cells (which is as far as I understand it perfectly sensible and a technical artifact due to the measurements themselves).

![SupFig15](https://user-images.githubusercontent.com/21232307/231721952-cd1bfbda-a4b7-48a7-b0eb-f9d93eae7f6f.png)
